core/services/vrf/v2/listener_v2_log_listener.go

```diff
@@ -0,0 +1,348 @@
+	// if we want to fulfill on time.
+	latestBlock, err := lp.LatestBlock()
+	if err != nil {
+		return nil, errors.Wrap(err, "LogPoller.LatestBlock()")
+	}
+	lsn.setLatestHead(latestBlock)
+
+	// We don't specify confs because each request can have a different conf above
+	// the minimum. So we do all conf handling in getConfirmedAt.
+	logs, err := lp.LogsWithSigs(
+		lastProcessedBlock,
+		latestBlock.BlockNumber,
+		[]common.Hash{lsn.coordinator.RandomWordsFulfilledTopic(), lsn.coordinator.RandomWordsRequestedTopic()},
+		lsn.coordinator.Address(),
+		pg.WithParentCtx(ctx),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "LogPoller.LogsWithSigs")
+	}
+
+	unfulfilled, unfulfilledLP, fulfilled := lsn.getUnfulfilled(logs, ll)
+	if len(unfulfilled) > 0 {
+		ll.Debugw("found unfulfilled logs", "unfulfilled", len(unfulfilled))
+	} else {
+		ll.Debugw("no unfulfilled logs found")
+	}
+
+	lsn.handleFulfilled(fulfilled)
+
+	return lsn.handleRequested(unfulfilled, unfulfilledLP, minConfs), nil
+}
+
+func (lsn *listenerV2) getUnfulfilled(logs []logpoller.Log, ll logger.Logger) (unfulfilled []RandomWordsRequested, unfulfilledLP []logpoller.Log, fulfilled map[string]RandomWordsFulfilled) {
+	var (
+		requested       = make(map[string]RandomWordsRequested)
+		requestedLP     = make(map[string]logpoller.Log)
+		errs            error
+		expectedKeyHash = lsn.job.VRFSpec.PublicKey.MustHash()
+	)
+	fulfilled = make(map[string]RandomWordsFulfilled)
+	for _, l := range logs {
+		if l.EventSig == lsn.coordinator.RandomWordsFulfilledTopic() {
+			parsed, err2 := lsn.coordinator.ParseRandomWordsFulfilled(l.ToGethLog())
+			if err2 != nil {
+				// should never happen
+				errs = multierr.Append(errs, err2)
+				continue
+			}
+			fulfilled[parsed.RequestID().String()] = parsed
+		} else if l.EventSig == lsn.coordinator.RandomWordsRequestedTopic() {
+			parsed, err2 := lsn.coordinator.ParseRandomWordsRequested(l.ToGethLog())
+			if err2 != nil {
+				// should never happen
+				errs = multierr.Append(errs, err2)
+				continue
+			}
+			keyHash := parsed.KeyHash()
+			if !bytes.Equal(keyHash[:], expectedKeyHash[:]) {
+				// wrong keyhash, can ignore
+				continue
+			}
+			requested[parsed.RequestID().String()] = parsed
+			requestedLP[parsed.RequestID().String()] = l
+		}
+	}
+	// should never happen, unsure if recoverable
+	// may be worth a panic
+	if errs != nil {
+		ll.Errorw("encountered parse errors", "err", errs)
+	}
+
+	if len(fulfilled) > 0 || len(requested) > 0 {
+		ll.Infow("found logs", "fulfilled", len(fulfilled), "requested", len(requested))
+	} else {
+		ll.Debugw("no logs found")
+	}
+
+	// find unfulfilled requests by comparing requested events with the fulfilled events
+	for reqID, req := range requested {
+		if _, isFulfilled := fulfilled[reqID]; !isFulfilled {
+			unfulfilled = append(unfulfilled, req)
+			unfulfilledLP = append(unfulfilledLP, requestedLP[reqID])
+		}
```

Comment:
The lists **unfulfilled** and **unfulfilledLP** are filtered purely based on reqID. Should we consider using a combination of **reqID, blockNum and blockHash** as key to identify reorg-ed requests with same reqID as new?

Reply:
I don't think so, IIUC these logs are a "snapshot" of the blockchain. You shouldn't see duplicated requests (i.e multiple logs w/ the same request ID but different BH/BN/etc.

Reply:
`getUnfulfilled` seems to be called with data from `evm.logs` table. Are duplicate/older logs for a given VRF requestID deleted from `evm.logs` during a re-org?

```
        // get randomness requested logs with the appropriate keyhash
	// keyhash is specified in topic1
	requests, err := lp.IndexedLogsCreatedAfter(
		lsn.coordinator.RandomWordsRequestedTopic(), // event sig
		lsn.coordinator.Address(),                   // address
		1,                                           // topic index
		[]common.Hash{lsn.job.VRFSpec.PublicKey.MustHash()}, // topic values
		fromTimestamp,       // from time
		logpoller.Finalized, // confs
	)
	if err != nil {
		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsRequested logs")
	}

	// fulfillments don't have keyhash indexed, we'll have to get all of them
	// TODO: can we instead write a single query that joins on request id's somehow?
	fulfillments, err := lp.LogsCreatedAfter(
		lsn.coordinator.RandomWordsFulfilledTopic(), // event sig
		lsn.coordinator.Address(),                   // address
		fromTimestamp,                               // from time
		logpoller.Finalized,                         // confs
	)
	if err != nil {
		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsFulfilled logs")
	}

	unfulfilled, _, _ := lsn.getUnfulfilled(append(requests, fulfillments...), ll)
```

Reply:
Yes, re-orged logs are deleted from the log poller table.